### PR TITLE
dynamic query restrictions

### PR DESCRIPTION
**Pull request by njr-11**
This PR contains some ideas that I was experimenting with locally for dynamic query restrictions #829 and hadn't been planning to propose until after a lot more discussion occurred.  However, because code changes are already being proposed and I wanted to illustrate how this should tie into the static metamodel, I decided to share this prematurely so everyone can see what I was originally aiming for with the pattern.

```
@Repository
public interface Products extends CrudRepository<Product, Long> {
    List<Product> search(String namePattern,
                         Restriction<Product> restriction,
                         Order<Product> order);
}
```

Usage with single restriction,
```
    List<Product> found = products.search(pattern,
                                          _Product.price.lessThan(50.0f),
                                          Order.by(_Product.price.desc(),
                                                   _Product.name.asc()));
```

Usage with multiple restrictions,
```
    List<Product> found = products.search(pattern,
                                          Restrict.all(_Product.price.greaterThan(25.0f),
                                                       _Product.price.lessThan(50.0f),
                                                       _Product.name.startsWith(pattern).ignoreCase(),
                                                       _Product.image.notNull()),
                                          Order.by(_Product.price.desc(),
                                                   _Product.name.asc()));
```

This was just the basics, and doesn't yet include not/negation, or Pattern/Range, which might make parts of it unnecessary.  It also just occurred to me that if we had a TextRestriction subclass, which I did not include in this, we could omit the `*IgnoreCase` methods and have `.ignoreCase()` instead.

### CODE REVIEW

**Comment by njr-11** [comment]
After a lot of difficulty reviewing another PR that combines several different new parts of Jakarta Data, I decided to just finish off updating this PR with how I expect Restrictions to work.

I was originally thinking that Restrictions would only come from the static metamodel, but others have expressed a desire that users be able to create them like they can for Sort.  I'm okay with that, but not by giving users access to record/implementation class constructors where they can populate them with any combinations of things they like.  With this PR, all implementation classes are made non-public and kept out of the API.  There are 2 places where users can obtain restrictions:

- static metamodel attributes (basic restrictions only)
- the Restrict class (both composite and basic restrictions)

This also aims to de-emphasize interfaces that the end user shouldn't care about (and only Data Providers need, such as Operator and the Composite restriction) out of the main view by placing them in inner classes.

There is no JavaDoc in this on purpose. The idea is to iterate over the design first, and later spend time on JavaDoc and other things.

Other spec function like Pattern, Range, Is, ... are intentionally kept out so we can focus only on getting Restriction correct.

`Restriction` could justifiably be placed in the jakarta.data package alongside some other special classes that are used as repository parameters (Sort, Order, Limit) although we might want to consider if it is worth it to have a separate `jakarta.data.restrict` package to separate it out like we do with `jakarta.data.page` for PageRequest, which is the other special parameter type. I think the latter is a good place for it, so that's where I put it in this PR.

greater/lessThan methods require a Comparable parameter (lots of things can be sortable, but I think they are all Comparable).  TextAttribute methods, and their equivalents on Restrict, require a String parameter.  I added a commented out method showing what it might be appropriate to do with Pattern parameter as well.  But ignore that if you don't want to think about Pattern yet.

I would like to see more error checking for values supplied to Restriction factory methods in general, and identify if we can use any more precise types that would help keep invalid objects out, but we can save that for when we are further along.

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (gavinking) [inline api/src/main/java/jakarta/data/restrict/Restrict.java:29]: Not sure I love how this breaks the pattern established by `Sort`/`Order`.

**Comment by gavinking** [comment]
> `Restriction` could justifiably be placed in the `jakarta.data` package

This is where I would have put it.

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (gavinking) [inline api/src/main/java/jakarta/data/restrict/Restrict.java:41]: Still no way to do `not()`, is that correct?

File: api/src/main/java/jakarta/data/metamodel/Attribute.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/metamodel/Attribute.java]: updated

File: api/src/main/java/jakarta/data/metamodel/Attribute.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/metamodel/Attribute.java]: Yes, we definitely want notIn and some others. I hadn't been thinking to include not* under this PR, but it's trivial to add, so I added it

File: api/src/main/java/jakarta/data/metamodel/Attribute.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/metamodel/Attribute.java]: added

File: api/src/main/java/jakarta/data/metamodel/SortableAttribute.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/metamodel/SortableAttribute.java:72]: added

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/restrict/Restrict.java]: Is the enum class the only problem here? If so, I can move the enum to another class and make Restrict into a normal class like Order is.  I didn't include this in the commit with the first round of fixes because I want to be sure what you are asking for.

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/restrict/Restrict.java]: Because it requires further discussion, I chose to defer it and leave a TODO comment (right after this) for it.  A follow-on PR can add it after we decide how best to do it.

**Review by njr-11 — COMMENTED** [review]
Added a commit to cover most of Gavin's review comments.
Also, I made an update to ensure ignoreCase is only available to String values, and not to numbers, dates, and so forth.

**Comment by njr-11** [comment]
> > `Restriction` could justifiably be placed in the `jakarta.data` package
> 
> This is where I would have put it.

That's fine. I have moved it to `jakarta.data` in the review fixes commit.

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/restrict/Restrict.java]: Yes, I think that's my preference. I don't think it's going to be normal for the user to want to access the enum values directly.

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/restrict/Restrict.java]: Yeah no worries, was just checking I had not missed it.

**Comment by gavinking** [comment]
> Added a commit to cover most of Gavin's review comments.

Thanks!

File: api/src/main/java/jakarta/data/restrict/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/restrict/Restrict.java]: I agree. The user will typically never look the composite type. The enum values are mainly for the Jakarta Data provider to check which type of composite it is. I made Restrict into a normal class and moved the enum elsewhere. Thanks for pointing this out.

**Comment by njr-11** [comment]
Switched methods such as startsWith(prefixPattern) to startsWith(prefix) and automatically add escaping to spare the user from that level of detail/complexity.

File: api/src/main/java/jakarta/data/Restriction.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/Restriction.java:47]: Change this one to: ignoreCase.

Comment (gavinking) [inline api/src/main/java/jakarta/data/Restriction.java:47]:
```suggestion
        boolean isCaseSensitive();
```

Comment (njr-11) [reply api/src/main/java/jakarta/data/Restriction.java:47]: Lacking a better name, we can use Gavin's suggestion of isCaseSensitive.  This method is here for the Jakarta Data provider rather than the user, and we can easily add a disclaimer in the JavaDoc that if a database is incapable of case sensitive comparison, it will ignore a value of true.

File: api/src/main/java/jakarta/data/TextRestrictionRecord.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/TextRestrictionRecord.java:24]: You can use `DefaultTextRestriction`

Comment (njr-11) [reply api/src/main/java/jakarta/data/TextRestrictionRecord.java:24]: I suppose another idea would be `TextRestrictionImpl`. That would make it very clear that it is a hidden internal class used for implementation.

Comment (otaviojava) [reply api/src/main/java/jakarta/data/TextRestrictionRecord.java:24]: I am OK with either `...Record`or `Default`, but I would avoid the `TextRestrictionImpl`

Comment (njr-11) [reply api/src/main/java/jakarta/data/TextRestrictionRecord.java:24]: Alright. `Default` seemed awkward to me, so I'll leave it as `TextRestrictionRecord` for now. We can rename it later if we think of something better.

File: api/src/main/java/jakarta/data/Restrict.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/Restrict.java:137]: It is not clear that `not` means, `notEquals´. 
 Please either, use `ne` or `NotEquals`

File: api/src/main/java/jakarta/data/Restrict.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/Restrict.java:145]: Can we remove all of those `no` method factory sequences?

It would be great if we have only two options:  by the method instead `negate`and a single method factory `not`.

It is pretty similar to the Java `Predicate`: https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/function/Predicate.html

File: api/src/main/java/jakarta/data/Restrict.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/Restrict.java:197]: Once everybody agreed with Gavin's `Pattern` class, why not centralize this behavior in that class?

File: api/src/main/java/jakarta/data/Restriction.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/Restriction.java:23]: Please remove this one from the inner class; it is good, mainly for the Javadoc.

File: api/src/main/java/jakarta/data/metamodel/SortableAttribute.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/metamodel/SortableAttribute.java:58]: Should this comparable be `Comparable<? extends Object> value`, I am not sure here.

File: api/src/main/java/jakarta/data/metamodel/TextAttribute.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/metamodel/TextAttribute.java:88]: The same point of using the Java core as reference to us:

https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/function/Predicate.html

File: api/src/main/java/jakarta/data/metamodel/SortableAttribute.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/metamodel/SortableAttribute.java:58]: I think it should probably just be `T`.

File: api/src/main/java/jakarta/data/Restrict.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/Restrict.java:145]: @otaviojava JDQL has operators like `not between`, `not like`, etc, so it makes sense to me for such things to be reflected here. And it's significantly less verbose.

File: api/src/main/java/jakarta/data/Restrict.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/Restrict.java:197]: Yes, I think it's much better to have a nice class which models this basic concept. It's a much semantically stronger construct.

File: api/src/main/java/jakarta/data/Restrict.java
Comment (otaviojava) [reply api/src/main/java/jakarta/data/Restrict.java:92]: Better!

File: api/src/main/java/jakarta/data/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/Restrict.java:145]: I agree with Gavin on this.

Being able to write code such as,

```
products.search(_Product.description.notContains("refurbished"));
```

is much more straightforward and readable than

```
products.search(_Product.description.contains("refurbished").negated());
```

or 

```
products.search(Restrict.not(_Product.description.contains("refurbished")));
```

File: api/src/main/java/jakarta/data/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/Restrict.java:197]: I agree, if Pattern gets added, this utility method can move to there.  Also note that the method is private, so the documentation on it is only for our own benefit and will not show in the API, so users won't even notice once it is moved.

File: api/src/main/java/jakarta/data/Restrict.java
Comment (njr-11) [reply api/src/main/java/jakarta/data/Restrict.java:137]: Sure, I can update that. Given that we have a `equalTo` method, the equivalent here will be `notEqualTo`

**Comment by njr-11** [comment]
Review comments have been addressed.

**Comment by otaviojava** [comment]
> Review comments have been addressed.

I saw, I am moving the documentation and the unit test that make sense against this PR.

I hope finish it today, thus, you will receive a PR soon.

**Comment by otaviojava** [comment]
@njr-11 done the first step: https://github.com/njr-11/data/pull/1

**Comment by njr-11** [comment]
Thanks for contributing the unit tests. That helps to confirm the correctness of the changes.
I was thinking maybe we should wait on adding JavaDoc until after this PR is merged, so that we don't need to rewrite it if anyone asks for further changes to the initial PR.

**Comment by otaviojava** [comment]
> Thanks for contributing the unit tests. That helps to confirm the correctness of the changes. I was thinking maybe we should wait on adding JavaDoc until after this PR is merged, so that we don't need to rewrite it if anyone asks for further changes to the initial PR.

It does make sense. 
I will focus on the Unit test to find some issues on it.

**Comment by otaviojava** [comment]
@njr-11 one more PR against yours.

**Comment by njr-11** [comment]
> @njr-11 one more PR against yours.

@otaviojava I don't see another PR against this anywhere.

**Comment by njr-11** [comment]
> > > @njr-11 one more PR against yours.
> > 
> > 
> > @otaviojava I don't see another PR against this anywhere.
> 
> This one here: [njr-11#2](https://github.com/njr-11/data/pull/2)

I see it now and have accepted your PR/commits for more unit tests.  Thanks for adding the test coverage.

**Comment by otaviojava** [comment]
> > > > @njr-11 one more PR against yours.
> > > 
> > > 
> > > @otaviojava I don't see another PR against this anywhere.
> > 
> > 
> > This one here: [njr-11#2](https://github.com/njr-11/data/pull/2)
> 
> I see it now and have accepted your PR/commits for more unit tests. Thanks for adding the test coverage.


@njr-11 

This is the last one:

https://github.com/njr-11/data/pull/3

I put the validation on field making the field mandatory

**Comment by otaviojava** [comment]
As a first step, IMHO, it is good to go.

@njr-11 @gavinking, what do you think?

**Review by otaviojava — APPROVED** [review]

**Pull request merged**
